site-src/geps/gep-851.md

```diff
@@ -0,0 +1,88 @@
+* Status: Implementable
+
+## TLDR
+
+Replace `CertificateRef` field with an `CertificateRefs` field in Gateway
+Listeners.
+
+## Goals
+
+* Allow certs to be attached for multiple unique Hostnames for the same
+  Listener. This is most helpful when the hostname is empty on the Listener and
+  attached Routes specify multiple hostnames.
+* Enable specifying RSA and ECDSA certs on the same Listener.
+* Improve cert upgrades by allowing new and old certs to be attached to the same
+  Listener.
+
+## Non-Goals
+
+* Be too prescriptive about how multiple certificates are handled for a Gateway
+  Listener. Each implementation may handle this slightly differently.
+
+## Introduction
+
+When we removed certificate references from HTTPRoute ([GEP
+746](/geps/gep-746/)) we lost the capability to attach multiple certificates to
+a Listener. This GEP proposes adding that capability back, directly to the
+Gateway Listener this time.
+
+In addition to helping with the self service approach, where app owners may
+attach multiple hostnames (and certs) to a single Listener, this can help with
+other use cases:
+
+* Attaching RSA and ECDSA certs.
+* Attaching both new and old certs during an upgrade.
+
+## API
+
+The `CertificateRef` field in `GatewayTLSConfig` would be replaced with the
+following `CertificateRefs` field:
+
+```go
+	// CertificateRefs contains a series of references to Kubernetes objects that
+	// contains TLS certificates and private keys. These certificates are used to
+	// establish a TLS handshake for requests that match the hostname of the
+	// associated listener.
+  //
+  // When multiple TLS certificates are specified, SNI matching MUST be used to
+  // determine the appropriate certificate to use.
+	//
+	// References to a resource in different namespace are invalid UNLESS there
+	// is a ReferencePolicy in the target namespace that allows the certificate
+	// to be attached. If a ReferencePolicy does not allow this reference, the
+	// "ResolvedRefs" condition MUST be set to False for this listener with the
+	// "InvalidCertificateRef" reason.
+	//
+	// This field is required to have at least one element when the mode is set to
+	// "Terminate" (default) and is optional otherwise.
+	//
+	// CertificateRefs can reference to standard Kubernetes resources, i.e.
+	// Secret, or implementation-specific custom resources.
+	//
+	// Support: Core (Kubernetes Secrets)
+	//
+	// Support: Implementation-specific (Other resource types)
+	//
+	// +optional
+	CertificateRefs []*SecretObjectReference `json:"certificateRefs,omitempty"`
+```
+
+## Self Service Model
+
+This change helps enable a self service model for TLS certificates. As an
+example, a Gateway may have a listener that does not specify a hostname. Two
+HTTP Routes may be attached, both with different hostnames. This change enables
+attaching different certificates for each of those hostnames to a Gateway
+Listener.
```

Comment:
Isn't it possible to enable the self-service model without this feature with listener collapsing and explicit per hostname listeners (with each of those specifying the TLS certificate)?

I'm a bit hesitant with this position because if the cluster-admin doesn't specify a hostname in the listener, they are leaving a wide hole in their infrastructure and trusting route owners with more than necessary for most organizations.
Listeners with an empty hostname are an escape hatch and are not meant for widespread usage. At least, that is my understanding.

Reply:
Yeah, this is a good point, maybe my example should be broader. Let's say the hostname on the Listener is `*.example.com` and we want to allow self service attachment of `store.staging.example.com` and `app.staging.example.com`. I think that's a reasonable use case for attaching certs to multiple subdomains.

Even that could be accomplished with multiple listeners, but that may not be something an infra admin wants to care about, they simply want to delegate everything under example.com to this set of namespaces/routes.

Reply:
> Let's say the hostname on the Listener is *.example.com and we want to allow self service attachment of store.staging.example.com and app.staging.example.com. I think that's a reasonable use case for attaching certs to multiple subdomains.

In my understanding, a single `*` implies one DNS label.
I _think_ our documentation notes the same: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/shared_types.go#L252-L254

Reply:
As per the current spec, the admin has two options:
- don't specify a hostname at all (I personally do not recommend this)
- specify hostnames: *.example.com *.staging.example.com - this is fairly common

Reply:
As I understand it, [Envoy can't restrict wildcard matching to a single DNS label](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-virtualhost). I know our docs were once more prescriptive in terms of a wildcard only matching a single DNS label, but I don't think they say that now. We should clarify this, but my interpretation of "prefixed with a single wildcard label" is that `*.example.com` is valid, but `*.*.example.com` is not. It does not actually limit what `*.example.com` should match.

/cc @howardjohn @jpeach @youngnick

Reply:
We discussed this in the meeting today, but to recap my understanding:
- it's important to distinguish here between the Gateway API rules for Gateway-Route attachment, and the way that those rules will be implemented.
- The Gateway API rules currently state that a `*` character only ever matches a single DNS label. So if the Listener has `*.example.com`, only `foo.example.com`, not `subdomain.foo.example.com` would match _if it was specified in the Route_.
- In terms of Envoy's _implementation_ of this, if the Route specifies `*.example.com` as the hostname, then when Envoy implements that, Envoy will match more than one label under the `*`. (I actually think that we should consider not allowing wildcards on Routes because of the combinatorial number of use-cases it allows, but not strongly enough to make it anything other than an aside).

Adding this feature means that there are effectively three places that a hostname needs to be matched (in the HTTPS case):
- SNI hostname via whatever is in all the certs. This means that most implementations will probably need separate routing chains per cert (Envoy certainly does).
- Listener Hostname, as we do now
- Route Hostname, as we do now.

For each hostname match, we allow either a precise or a wildcard match, or the Listener and Route hostnames can be unspecified. So, our number of combinations is 2 x 3 x 3 = 18 combinations that we need to define rules on, for when things line up. If hostnames are precise but don't all match, then that's a separate degree of freedom. What I'm saying by this paragraph is that, in adding this feature, we have to carefully define the rules for what supersedes what, and what is valid.

To choose some examples out of the space:
- For a cert that handles `*.example.com`, a Listener hostname of `foo.example.com`, and a Route of `*.example.com`, everything will attach, but the actual routing config must necessarily be limited to `foo.example.com`.
- For a cert that handles `foo.example.com`, a Listener hostname of `*.example.com`, and a Route of `bar.example.com`, obviously, no match.
- For a cert that handles `foo.example.com`, a LIstener hostname of `*.example.com`, and a Route of `foo.example.com`, we have a match, and the routing chain must also be `foo.example.com`.
- The case we talked about today is when there is a cert that handles `*.example.com`, and the Listener or Route hostnames are either unset or also `*.example.com`, then the eventual routing chain that's created will do SNI for *.example.com, and then Envoy will route host headers for `*.example.com`, which could conceivably include multiple levels of routing. However, without having the SNI and Host header being different, this won't be a problem - this is an example of why we should ensure that the SNI and Host header must match.

Maybe the shortest way to sy what I think the rules should be are:
- wildcard matches are only ever one label in the Gateway API.
- If wildcard matches and precise matches are being combined, the precise one is the result (as long as they match).
- In general you should try to keep wildcard matches at the Gateway level and leave Route hostname matches for precise matches.

Reply:
Thanks for taking the time to write this up @youngnick! This is a really great summary of the levels of complexity this change would add. As I've been reading through this and other comments here, I'm wondering if it makes sense to limit the scope of this change to support for RSA and EC certs. Although much smaller than the original intent, this accomplishes several important things:

- Supports attaching both RSA and EC certs to the same listener
- Transitions CertRefs to a list, allowing future expansion of this field simply by loosening validation

The vast majority of the complexity in this discussion is based on unspecified future work to enable self service certificate attachment. It would likely make more sense to wait until we were ready to fully tackle that before we determine how this particular use of multiple certs should be handled.